File: addons/metrics.py
"""
metrics.py - Native mitmproxy addon for metrics collection

Collects per-domain statistics:
- Request counts, success rates
- Latency tracking
- Error categorization
- Block counts by source

Exposes metrics via admin API in JSON and Prometheus formats.

Usage:
    mitmdump -s addons/metrics.py
"""

import logging
import threading
import time
from dataclasses import dataclass

from mitmproxy import http

log = logging.getLogger("safeyolo.metrics")


@dataclass
class DomainStats:
    """Per-domain statistics."""
    requests: int = 0
    successes: int = 0
    blocked_credential: int = 0
    blocked_yara: int = 0
    blocked_pattern: int = 0
    blocked_injection: int = 0
    upstream_429s: int = 0
    upstream_5xx: int = 0
    timeouts: int = 0

    latency_sum_ms: float = 0
    latency_count: int = 0
    latency_max_ms: float = 0

    @property
    def success_rate(self) -> float:
        if self.requests == 0:
            return 1.0
        return self.successes / self.requests

    @property
    def avg_latency_ms(self) -> float:
        if self.latency_count == 0:
            return 0
        return self.latency_sum_ms / self.latency_count

    def to_dict(self) -> dict:
        return {
            "requests": self.requests,
            "successes": self.successes,
            "success_rate": round(self.success_rate, 3),
            "blocked": {
                "credential": self.blocked_credential,
                "yara": self.blocked_yara,
                "pattern": self.blocked_pattern,
                "injection": self.blocked_injection,
            },
            "upstream_errors": {
                "429s": self.upstream_429s,
                "5xx": self.upstream_5xx,
                "timeouts": self.timeouts,
            },
            "latency_ms": {
                "avg": round(self.avg_latency_ms, 1),
                "max": round(self.latency_max_ms, 1),
            },
        }


class MetricsCollector:
    """
    Native mitmproxy addon for metrics collection.

    Observes all requests/responses and tracks statistics per domain.
    """

    name = "metrics"

    def __init__(self):
        self._lock = threading.Lock()
        self._start_time = time.time()
        self._domain_stats: dict[str, DomainStats] = {}

        # Global counters
        self.requests_total = 0
        self.requests_success = 0
        self.requests_blocked = 0
        self.requests_error = 0

    def _get_domain_stats(self, domain: str) -> DomainStats:
        with self._lock:
            if domain not in self._domain_stats:
                self._domain_stats[domain] = DomainStats()
            return self._domain_stats[domain]

    def request(self, flow: http.HTTPFlow):
        """Record request."""
        self.requests_total += 1
        self._get_domain_stats(flow.request.host).requests += 1
        flow.metadata["metrics_start_time"] = time.time()

    def response(self, flow: http.HTTPFlow):
        """Record response metrics."""
        domain = flow.request.host
        stats = self._get_domain_stats(domain)

        # Calculate latency
        start_time = flow.metadata.get("metrics_start_time")
        latency_ms = 0.0
        if start_time:
            latency_ms = (time.time() - start_time) * 1000

        # Check if blocked
        blocked_by = flow.metadata.get("blocked_by")
        if blocked_by:
            self.requests_blocked += 1
            if blocked_by == "credential-guard":
                stats.blocked_credential += 1
            elif blocked_by == "yara-scanner":
                stats.blocked_yara += 1
            elif blocked_by == "pattern-scanner":
                stats.blocked_pattern += 1
            elif blocked_by == "prompt-injection":
                stats.blocked_injection += 1
            return

        if not flow.response:
            return

        # Check response status
        status = flow.response.status_code

        if status == 429:
            stats.upstream_429s += 1
        elif status >= 500:
            stats.upstream_5xx += 1
            if status == 504:
                stats.timeouts += 1
                self.requests_error += 1
        elif status < 400:
            self.requests_success += 1
            stats.successes += 1
            stats.latency_sum_ms += latency_ms
            stats.latency_count += 1
            stats.latency_max_ms = max(stats.latency_max_ms, latency_ms)

    def get_json(self) -> dict:
        """Get metrics as JSON."""
        uptime = time.time() - self._start_time

        # Copy under lock for thread safety
        with self._lock:
            domain_stats_copy = dict(self._domain_stats)

        # Sort domains by request count
        sorted_domains = sorted(
            domain_stats_copy.items(),
            key=lambda x: x[1].requests,
            reverse=True,
        )

        # Identify problem domains
        problem_domains = []
        for domain, stats in sorted_domains:
            issues = []
            if stats.success_rate < 0.9 and stats.requests > 10:
                issues.append(f"low_success_rate:{stats.success_rate:.1%}")
            if stats.upstream_429s > 5:
                issues.append(f"upstream_429s:{stats.upstream_429s}")
            if issues:
                problem_domains.append({"domain": domain, "issues": issues})

        return {
            "uptime_seconds": round(uptime, 1),
            "summary": {
                "requests_total": self.requests_total,
                "requests_success": self.requests_success,
                "requests_blocked": self.requests_blocked,
                "requests_error": self.requests_error,
                "success_rate": round(
                    self.requests_success / max(1, self.requests_total), 3
                ),
                "domains_tracked": len(domain_stats_copy),
            },
            "problem_domains": problem_domains,
            "domains": {
                domain: stats.to_dict() for domain, stats in sorted_domains[:20]
            },
        }

    def get_prometheus(self) -> str:
        """Get metrics in Prometheus format."""
        lines = []
        uptime = time.time() - self._start_time

        # Copy under lock for thread safety
        with self._lock:
            domain_stats_copy = dict(self._domain_stats)

        # Global metrics
        lines.append("# HELP safeyolo_uptime_seconds Proxy uptime")
        lines.append("# TYPE safeyolo_uptime_seconds gauge")
        lines.append(f"safeyolo_uptime_seconds {uptime:.1f}")

        lines.append("# HELP safeyolo_requests_total Total requests")
        lines.append("# TYPE safeyolo_requests_total counter")
        lines.append(f"safeyolo_requests_total {self.requests_total}")

        lines.append("# HELP safeyolo_requests_success Successful requests")
        lines.append("# TYPE safeyolo_requests_success counter")
        lines.append(f"safeyolo_requests_success {self.requests_success}")

        lines.append("# HELP safeyolo_requests_blocked Blocked requests")
        lines.append("# TYPE safeyolo_requests_blocked counter")
        lines.append(f"safeyolo_requests_blocked {self.requests_blocked}")

        # Per-domain metrics
        lines.append("")
        lines.append("# HELP safeyolo_domain_requests_total Requests per domain")
        lines.append("# TYPE safeyolo_domain_requests_total counter")
        for domain, stats in domain_stats_copy.items():
            lines.append(f'safeyolo_domain_requests_total{{domain="{domain}"}} {stats.requests}')

        lines.append("")
        lines.append("# HELP safeyolo_domain_success_rate Success rate per domain")
        lines.append("# TYPE safeyolo_domain_success_rate gauge")
        for domain, stats in domain_stats_copy.items():
            lines.append(f'safeyolo_domain_success_rate{{domain="{domain}"}} {stats.success_rate:.3f}')

        lines.append("")
        lines.append("# HELP safeyolo_domain_latency_avg_ms Average latency per domain")
        lines.append("# TYPE safeyolo_domain_latency_avg_ms gauge")
        for domain, stats in domain_stats_copy.items():
            lines.append(f'safeyolo_domain_latency_avg_ms{{domain="{domain}"}} {stats.avg_latency_ms:.1f}')

        return "\n".join(lines) + "\n"

    def get_stats(self) -> dict:
        """Get basic stats for admin API."""
        with self._lock:
            domains_tracked = len(self._domain_stats)
        return {
            "requests_total": self.requests_total,
            "requests_success": self.requests_success,
            "requests_blocked": self.requests_blocked,
            "domains_tracked": domains_tracked,
        }


# mitmproxy addon instance
addons = [MetricsCollector()]

File: proxy/src/metrics.rs
//! Reached HTTP metrics and source report formats. The runtime owns one shared
//! collector and retains the returned request timestamp per exchange.

use std::{
    fmt,
    sync::{Mutex, MutexGuard},
};

use indexmap::IndexMap;
use num_bigint::BigInt;
use serde_json::Value;
use zeroize::Zeroize;

use crate::{
    circuits::{CircuitValue as C, ErrorKind as NumericError},
    network_guard::sanitize_with_limit,
};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ErrorKind {
    Poisoned,
    Type,
    Overflow,
    Compatibility,
}
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Error(ErrorKind);
impl Error {
    pub fn kind(self) -> ErrorKind {
        self.0
    }
}
impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("metrics operation failed")
    }
}
impl std::error::Error for Error {}
pub type Result<T> = std::result::Result<T, Error>;

#[derive(Default)]
struct Domain {
    requests: BigInt,
    successes: BigInt,
    credential: BigInt,
    yara: BigInt,
    pattern: BigInt,
    injection: BigInt,
    upstream_429: BigInt,
    upstream_5xx: BigInt,
    timeouts: BigInt,
    // None is the source's initial integer 0, despite its float annotations.
    latency_sum: Option<f64>,
    latency_count: BigInt,
    latency_max: Option<f64>,
}
impl Domain {
    fn success_rate(&self) -> Result<f64> {
        if self.requests == BigInt::from(0) {
            Ok(1.0)
        } else {
            ratio(&self.successes, &self.requests)
        }
    }
    fn average(&self) -> Result<C> {
        if self.latency_count == BigInt::from(0) {
            Ok(0.into())
        } else {
            divide(
                C::Float(self.latency_sum.unwrap_or(0.0)),
                &self.latency_count,
            )
        }
    }
    fn document(&self) -> Result<C> {
        Ok(object([
            ("requests", self.requests.clone().into()),
            ("successes", self.successes.clone().into()),
            ("success_rate", C::Float(rounded(self.success_rate()?, 3))),
            (
                "blocked",
                object([
                    ("credential", self.credential.clone().into()),
                    ("yara", self.yara.clone().into()),
                    ("pattern", self.pattern.clone().into()),
                    ("injection", self.injection.clone().into()),
                ]),
            ),
            (
                "upstream_errors",
                object([
                    ("429s", self.upstream_429.clone().into()),
                    ("5xx", self.upstream_5xx.clone().into()),
                    ("timeouts", self.timeouts.clone().into()),
                ]),
            ),
            (
                "latency_ms",
                object([
                    ("avg", round_value(self.average()?, 1)),
                    (
                        "max",
                        self.latency_max
                            .map_or(0.into(), |n| C::Float(rounded(n, 1))),
                    ),
                ]),
            ),
        ]))
    }
}

struct State {
    started: f64,
    total: BigInt,
    success: BigInt,
    blocked: BigInt,
    error: BigInt,
    sources: IndexMap<String, BigInt>,
    domains: IndexMap<String, Domain>,
}
impl Drop for State {
    fn drop(&mut self) {
        for (mut name, _) in self.sources.drain(..) {
            name.zeroize();
        }
        for (mut name, _) in self.domains.drain(..) {
            name.zeroize();
        }
    }
}

pub struct Metrics {
    state: Mutex<State>,
}
impl Metrics {
    /// Sample the source uptime origin once at process-owner creation.
    pub fn new(clock: impl FnOnce() -> f64) -> Self {
        Self {
            state: Mutex::new(State {
                started: clock(),
                total: 0.into(),
                success: 0.into(),
                blocked: 0.into(),
                error: 0.into(),
                sources: IndexMap::new(),
                domains: IndexMap::new(),
            }),
        }
    }
    fn lock(&self) -> Result<MutexGuard<'_, State>> {
        self.state.lock().map_err(|_| Error(ErrorKind::Poisoned))
    }

    /// Called only when the late source request hook is reached. The returned
    /// clock value replaces this exchange's previous metrics_start_time.
    pub fn request(&self, host: &str, clock: impl FnOnce() -> f64) -> Result<f64> {
        let mut state = self.lock()?;
        state.total += 1;
        state.domains.entry(host.into()).or_default().requests += 1;
        drop(state);
        Ok(clock())
    }

    /// This observes the validated response hook, not transport errors or
    /// bytes delivered to a client. A response can precede its request hook.
    pub fn response(
        &self,
        host: &str,
        start: Option<f64>,
        blocked_by: Option<&C>,
        status: Option<u16>,
        clock: impl FnOnce() -> f64,
    ) -> Result<()> {
        let mut state = self.lock()?;
        state.domains.entry(host.into()).or_default();
        drop(state);
        let elapsed = start
            .filter(|start| *start != 0.0)
            .map_or(0.0, |start| (clock() - start) * 1000.0);
        let mut state = self.lock()?;
        if let Some(blocked) = blocked_by.filter(|value| value.truthy()) {
            state.blocked += 1;
            let source = match blocked {
                C::Other(Value::String(source)) => source,
                C::Array(_) | C::Object(_) | C::Other(Value::Array(_) | Value::Object(_)) => {
                    return Err(Error(ErrorKind::Type));
                }
                // Source hashable nonstring keys are not produced by the live
                // native pipeline. Preserve reached partial effects and expose
                // the representation gap instead of coercing a block source.
                _ => return Err(Error(ErrorKind::Compatibility)),
            };
            let stats = state.domains.get_mut(host).expect("inserted domain");
            match source.as_str() {
                "credential-guard" => stats.credential += 1,
                "yara-scanner" => stats.yara += 1,
                "pattern-scanner" => stats.pattern += 1,
                "prompt-injection" => stats.injection += 1,
                _ => {}
            }
            *state.sources.entry(source.clone()).or_default() += 1;
            return Ok(());
        }
        let Some(status) = status else {
            return Ok(());
        };
        if status == 429 {
            state.domains.get_mut(host).unwrap().upstream_429 += 1;
        } else if status >= 500 {
            let stats = state.domains.get_mut(host).unwrap();
            stats.upstream_5xx += 1;
            if status == 504 {
                stats.timeouts += 1;
                state.error += 1;
            }
        } else if status < 400 {
            state.success += 1;
            let stats = state.domains.get_mut(host).unwrap();
            stats.successes += 1;
            stats.latency_sum = Some(stats.latency_sum.unwrap_or(0.0) + elapsed);
            stats.latency_count += 1;
            // Python max(old, new) retains the first operand on equal/NaN.
            if elapsed > stats.latency_max.unwrap_or(0.0) {
                stats.latency_max = Some(elapsed);
            }
        }
        Ok(())
    }

    /// Source's basic operator report does not sample the clock.
    pub fn get_stats(&self) -> Result<C> {
        let state = self.lock()?;
        Ok(object([
            ("requests_total", state.total.clone().into()),
            ("requests_success", state.success.clone().into()),
            ("requests_blocked", state.blocked.clone().into()),
            ("blocks_by_source", sources(&state)),
            ("domains_tracked", BigInt::from(state.domains.len()).into()),
        ]))
    }

    /// Full source JSON rendering remains available to an explicit caller;
    /// this component installs no HTTP endpoint.
    pub fn get_json(&self, clock: impl FnOnce() -> f64) -> Result<C> {
        let now = clock();
        let state = self.lock()?;
        let mut domains: Vec<_> = state.domains.iter().collect();
        domains.sort_by(|left, right| right.1.requests.cmp(&left.1.requests));
        let mut problems = Vec::new();
        for (host, stats) in &domains {
            let mut issues = Vec::new();
            let rate = stats.success_rate()?;
            if rate < 0.9 && stats.requests > BigInt::from(10) {
                issues.push(text(&format!(
                    "low_success_rate:{}%",
                    fixed(rate * 100.0, 1)
                )));
            }
            if stats.upstream_429 > BigInt::from(5) {
                issues.push(text(&format!("upstream_429s:{}", stats.upstream_429)));
            }
            if !issues.is_empty() {
                problems.push(object([
                    ("domain", text(host)),
                    ("issues", C::Array(issues)),
                ]));
            }
        }
        let denominator = state.total.clone().max(1.into());
        Ok(object([
            ("uptime_seconds", C::Float(rounded(now - state.started, 1))),
            (
                "summary",
                object([
                    ("requests_total", state.total.clone().into()),
                    ("requests_success", state.success.clone().into()),
                    ("requests_blocked", state.blocked.clone().into()),
                    ("requests_error", state.error.clone().into()),
                    ("blocks_by_source", sources(&state)),
                    (
                        "success_rate",
                        C::Float(rounded(ratio(&state.success, &denominator)?, 3)),
                    ),
                    ("domains_tracked", BigInt::from(state.domains.len()).into()),
                ]),
            ),
            ("problem_domains", C::Array(problems)),
            (
                "domains",
                C::Object(
                    domains
                        .into_iter()
                        .take(20)
                        .map(|(host, stats)| Ok((host.clone(), stats.document()?)))
                        .collect::<Result<_>>()?,
                ),
            ),
        ]))
    }

    pub fn get_prometheus(&self, clock: impl FnOnce() -> f64) -> Result<String> {
        let now = clock();
        let state = self.lock()?;
        let mut lines = vec![
            "# HELP safeyolo_uptime_seconds Proxy uptime".into(),
            "# TYPE safeyolo_uptime_seconds gauge".into(),
            format!("safeyolo_uptime_seconds {}", fixed(now - state.started, 1)),
            "# HELP safeyolo_requests_total Total requests".into(),
            "# TYPE safeyolo_requests_total counter".into(),
            format!("safeyolo_requests_total {}", state.total),
            "# HELP safeyolo_requests_success Successful requests".into(),
            "# TYPE safeyolo_requests_success counter".into(),
            format!("safeyolo_requests_success {}", state.success),
            "# HELP safeyolo_requests_blocked Blocked requests".into(),
            "# TYPE safeyolo_requests_blocked counter".into(),
            format!("safeyolo_requests_blocked {}", state.blocked),
        ];
        if !state.sources.is_empty() {
            lines.extend([
                String::new(),
                "# HELP safeyolo_blocks_by_source Blocked requests by addon source".into(),
                "# TYPE safeyolo_blocks_by_source counter".into(),
            ]);
            let mut sources: Vec<_> = state.sources.iter().collect();
            sources.sort_by(|left, right| left.0.cmp(right.0));
            for (source, count) in sources {
                lines.push(format!(
                    "safeyolo_blocks_by_source{{source=\"{}\"}} {count}",
                    sanitize_with_limit(source, 64)
                ));
            }
        }
        lines.extend([
            String::new(),
            "# HELP safeyolo_domain_requests_total Requests per domain".into(),
            "# TYPE safeyolo_domain_requests_total counter".into(),
        ]);
        for (host, stats) in &state.domains {
            lines.push(format!(
                "safeyolo_domain_requests_total{{domain=\"{}\"}} {}",
                sanitize_with_limit(host, 253),
                stats.requests
            ));
        }
        lines.extend([
            String::new(),
            "# HELP safeyolo_domain_success_rate Success rate per domain".into(),
            "# TYPE safeyolo_domain_success_rate gauge".into(),
        ]);
        for (host, stats) in &state.domains {
            lines.push(format!(
                "safeyolo_domain_success_rate{{domain=\"{}\"}} {}",
                sanitize_with_limit(host, 253),
                fixed(stats.success_rate()?, 3)
            ));
        }
        lines.extend([
            String::new(),
            "# HELP safeyolo_domain_latency_avg_ms Average latency per domain".into(),
            "# TYPE safeyolo_domain_latency_avg_ms gauge".into(),
        ]);
        for (host, stats) in &state.domains {
            let value = match stats.average()? {
                C::Float(value) => value,
                _ => 0.0,
            };
            lines.push(format!(
                "safeyolo_domain_latency_avg_ms{{domain=\"{}\"}} {}",
                sanitize_with_limit(host, 253),
                fixed(value, 1)
            ));
        }
        Ok(lines.join("\n") + "\n")
    }
}

fn divide(numerator: C, denominator: &BigInt) -> Result<C> {
    numerator
        .divide(&C::Integer(denominator.clone()))
        .map_err(|error| {
            Error(match error.kind() {
                NumericError::Type => ErrorKind::Type,
                NumericError::Overflow => ErrorKind::Overflow,
                _ => ErrorKind::Compatibility,
            })
        })
}
fn ratio(numerator: &BigInt, denominator: &BigInt) -> Result<f64> {
    let C::Float(value) = divide(numerator.clone().into(), denominator)? else {
        unreachable!("true division")
    };
    Ok(value)
}
fn rounded(value: f64, places: usize) -> f64 {
    if value.is_finite() {
        format!("{value:.places$}")
            .parse()
            .expect("formatted float")
    } else {
        value
    }
}
fn round_value(value: C, places: usize) -> C {
    if let C::Float(value) = value {
        C::Float(rounded(value, places))
    } else {
        value
    }
}
fn fixed(value: f64, places: usize) -> String {
    if value.is_nan() {
        "nan".into()
    } else {
        format!("{value:.places$}")
    }
}
fn text(value: &str) -> C {
    C::Other(Value::String(value.into()))
}
fn object<const N: usize>(fields: [(&str, C); N]) -> C {
    C::Object(
        fields
            .into_iter()
            .map(|(key, value)| (key.into(), value))
            .collect(),
    )
}
fn sources(state: &State) -> C {
    C::Object(
        state
            .sources
            .iter()
            .map(|(key, count)| (key.clone(), count.clone().into()))
            .collect(),
    )
}

#[cfg(test)]
mod tests;
